kafka-client/src/main/java/dev/responsive/kafka/api/stores/TtlProvider.java

```diff
@@ -0,0 +1,295 @@
+      final Serde<K> keySerde
+  ) {
+    if (ttlType.equals(TtlType.VALUE) || ttlType.equals(TtlType.KEY_AND_VALUE)) {
+      throw new IllegalArgumentException("Must choose only key, value, or key-and-value ttl");
+    }
+
+    if (keySerde == null || keySerde.deserializer() == null) {
+      throw new IllegalArgumentException("The key Serde and Deserializer must not be null");
+    }
+
+    return new TtlProvider<>(
+        TtlType.KEY,
+        defaultTtl,
+        (k, ignored) -> computeTtlFromKey.apply(k),
+        keySerde,
+        null
+    );
+  }
+
+  /**
+   * @return the same TtlProvider with a value-based override function
+   */
+  public TtlProvider<K, V> fromValue(
+      final Function<V, Optional<TtlDuration>> computeTtlFromValue,
+      final Serde<V> valueSerde
+  ) {
+    if (ttlType.equals(TtlType.KEY) || ttlType.equals(TtlType.KEY_AND_VALUE)) {
+      throw new IllegalArgumentException("Must choose only key, value, or key-and-value ttl");
+    }
+
+    if (valueSerde == null || valueSerde.deserializer() == null) {
+      throw new IllegalArgumentException("The value Serde and Deserializer must not be null");
+    }
+
+    return new TtlProvider<>(
+        TtlType.VALUE,
+        defaultTtl,
+        (ignored, v) -> computeTtlFromValue.apply(v),
+        null,
+        valueSerde);
+  }
+
+  /**
+   * @return the same TtlProvider with a key-and-value-based override function
+   */
+  public TtlProvider<K, V> fromKeyAndValue(
+      final BiFunction<K, V, Optional<TtlDuration>> computeTtlFromKeyAndValue,
+      final Serde<K> keySerde,
+      final Serde<V> valueSerde
+  ) {
+    if (ttlType.equals(TtlType.KEY) || ttlType.equals(TtlType.VALUE)) {
+      throw new IllegalArgumentException("Must choose only key, value, or key-and-value ttl");
+    }
+
+    if (keySerde == null || keySerde.deserializer() == null) {
+      throw new IllegalArgumentException("The key Serde and Deserializer must not be null");
+    } else if (valueSerde == null || valueSerde.deserializer() == null) {
+      throw new IllegalArgumentException("The value Serde and Deserializer must not be null");
+    }
+
+    return new TtlProvider<>(
+        TtlType.KEY_AND_VALUE,
+        defaultTtl,
+        computeTtlFromKeyAndValue,
+        keySerde,
+        valueSerde
+    );
+  }
+
+  public static class TtlDuration {
+
+    public enum Ttl {
+      INFINITE,
+      FINITE
+    }
+
+    public static TtlDuration of(final Duration ttl) {
+      if (ttl.compareTo(Duration.ZERO) <= 0) {
+        throw new IllegalArgumentException("ttl duration must be greater than zero");
+      }
+      return new TtlDuration(ttl, Ttl.FINITE);
+    }
+
+    // No ttl will be applied, in other words infinite retention
+    public static TtlDuration noTtl() {
+      return new TtlDuration(Duration.ZERO, Ttl.INFINITE);
+    }
+
+    // TODO(sophie): store ttl as long to avoid Duration conversions on the hot path
```

Comment:
@agavra I'm actually a bit concerned about this. Even if we store it as millis we'll still have a Duration on the hot path since this is the return type. What if we just bite the bullet and have the user return a TTL in terms of millis rather than a Duration?

I know it's not as nice API-wise but To be honest I think most Streams users are probably pretty used to reasoning about time in terms of milliseconds. So I'm personally leaning towards replacing Duration with long millis entirely. Thoughts?

Reply:
Do we have evidence that this is actually significant on a per-retrieved doc basis? 

A few bitwise CPU operations seems to be dwarfed by the potential network call + the re-serialization in my opinion. Even if hits the in-memory cache, we're talking about loads from main memory into L1 cache.

The cost of creating a `Duration` is ~15ns (thanks ChatGPT for the estimation, looks legit to me, and half of that cost is just object allocation) and the cost of calling `toMillis` is ~1ns (assuming there's no nanosecond component), so we're talking 16ns per-key returned overhead for using Duration on a 3.2Ghz CPU.